<!DOCTYPE html>
<html>


<!-- Mirrored from matteo.vaccari.name/blog/archives/875 by HTTrack Website Copier/3.x [XR&CO'2014], Fri, 24 Sep 2021 10:25:15 GMT -->
<!-- Added by HTTrack --><meta http-equiv="content-type" content="text/html;charset=UTF-8" /><!-- /Added by HTTrack -->
<head>
<meta charset="UTF-8" />
<meta name="viewport" content="initial-scale=1">

<title>Extreme Enthusiasm  &raquo; Blog Archive   &raquo; Notes on exception handling</title>

<meta name="generator" content="WordPress 5.1.1" /> <!-- leave this for stats -->

<link rel="stylesheet" href="../wp-content/themes/mv2/style.css" type="text/css" media="screen" />
<link rel="alternate" type="application/rss+xml" title="Extreme Enthusiasm RSS Feed" href="../feed" />
<link rel="pingback" href="../xmlrpc.php" />

<link rel='dns-prefetch' href='http://s.w.org/' />
<link rel="alternate" type="application/rss+xml" title="Extreme Enthusiasm &raquo; Notes on exception handling Comments Feed" href="875/feed" />
		<script type="text/javascript">
			window._wpemojiSettings = {"baseUrl":"https:\/\/s.w.org\/images\/core\/emoji\/11.2.0\/72x72\/","ext":".png","svgUrl":"https:\/\/s.w.org\/images\/core\/emoji\/11.2.0\/svg\/","svgExt":".svg","source":{"concatemoji":"http:\/\/matteo.vaccari.name\/blog\/wp-includes\/js\/wp-emoji-release.min.js?ver=5.1.1"}};
			!function(a,b,c){function d(a,b){var c=String.fromCharCode;l.clearRect(0,0,k.width,k.height),l.fillText(c.apply(this,a),0,0);var d=k.toDataURL();l.clearRect(0,0,k.width,k.height),l.fillText(c.apply(this,b),0,0);var e=k.toDataURL();return d===e}function e(a){var b;if(!l||!l.fillText)return!1;switch(l.textBaseline="top",l.font="600 32px Arial",a){case"flag":return!(b=d([55356,56826,55356,56819],[55356,56826,8203,55356,56819]))&&(b=d([55356,57332,56128,56423,56128,56418,56128,56421,56128,56430,56128,56423,56128,56447],[55356,57332,8203,56128,56423,8203,56128,56418,8203,56128,56421,8203,56128,56430,8203,56128,56423,8203,56128,56447]),!b);case"emoji":return b=d([55358,56760,9792,65039],[55358,56760,8203,9792,65039]),!b}return!1}function f(a){var c=b.createElement("script");c.src=a,c.defer=c.type="text/javascript",b.getElementsByTagName("head")[0].appendChild(c)}var g,h,i,j,k=b.createElement("canvas"),l=k.getContext&&k.getContext("2d");for(j=Array("flag","emoji"),c.supports={everything:!0,everythingExceptFlag:!0},i=0;i<j.length;i++)c.supports[j[i]]=e(j[i]),c.supports.everything=c.supports.everything&&c.supports[j[i]],"flag"!==j[i]&&(c.supports.everythingExceptFlag=c.supports.everythingExceptFlag&&c.supports[j[i]]);c.supports.everythingExceptFlag=c.supports.everythingExceptFlag&&!c.supports.flag,c.DOMReady=!1,c.readyCallback=function(){c.DOMReady=!0},c.supports.everything||(h=function(){c.readyCallback()},b.addEventListener?(b.addEventListener("DOMContentLoaded",h,!1),a.addEventListener("load",h,!1)):(a.attachEvent("onload",h),b.attachEvent("onreadystatechange",function(){"complete"===b.readyState&&c.readyCallback()})),g=c.source||{},g.concatemoji?f(g.concatemoji):g.wpemoji&&g.twemoji&&(f(g.twemoji),f(g.wpemoji)))}(window,document,window._wpemojiSettings);
		</script>
		<style type="text/css">
img.wp-smiley,
img.emoji {
	display: inline !important;
	border: none !important;
	box-shadow: none !important;
	height: 1em !important;
	width: 1em !important;
	margin: 0 .07em !important;
	vertical-align: -0.1em !important;
	background: none !important;
	padding: 0 !important;
}
</style>
	<link rel='stylesheet' id='wp-block-library-css'  href='../wp-includes/css/dist/block-library/style.min3c21.css?ver=5.1.1' type='text/css' media='all' />
<link rel='https://api.w.org/' href='../wp-json/index.html' />
<link rel="EditURI" type="application/rsd+xml" title="RSD" href="../xmlrpc0db0.php?rsd" />
<link rel="wlwmanifest" type="application/wlwmanifest+xml" href="../wp-includes/wlwmanifest.xml" /> 
<link rel='prev' title='How to host a JavaScript coding dojo' href='868.html' />
<link rel='next' title='Il mio corso di TDD' href='886.html' />
<meta name="generator" content="WordPress 5.1.1" />
<link rel="canonical" href="875.html" />
<link rel='shortlink' href='../index527a.html?p=875' />
<link rel="alternate" type="application/json+oembed" href="../wp-json/oembed/1.0/embed893e.json?url=http%3A%2F%2Fmatteo.vaccari.name%2Fblog%2Farchives%2F875" />
<link rel="alternate" type="text/xml+oembed" href="../wp-json/oembed/1.0/embed43f9?url=http%3A%2F%2Fmatteo.vaccari.name%2Fblog%2Farchives%2F875&amp;format=xml" />
</head>
<body>
<div id="page">

<div id="header">
	<h1><a href="../../blog">Extreme Enthusiasm</a></h1>
</div>

<div id='maincontent'>
    <div id='main-col'>

  
		<div class="navigation">
			<div class="alignleft">&laquo; <a href="868.html" rel="prev">How to host a JavaScript coding dojo</a></div>
			<div class="alignright"><a href="886.html" rel="next">Il mio corso di TDD</a> &raquo;</div>
		</div>

		<div class="post" id="post-875">
			<h2><a href="875.html" rel="bookmark" title="Permanent Link: Notes on exception handling">Notes on exception handling</a></h2>

			<div class="entrytext">
				<style>
  code {
    font-size: smaller;
  }
  pre {
    page-break-inside: avoid;
  }
</style>
<blockquote>
<p>  If a function be advertised to return an error code in the event of difficulties, thou shalt check for that code, yea, even though the checks triple the size of thy code and produce aches in thy typing fingers, for if thou thinkest &#8220;it cannot happen to me&#8221;, the gods shall surely punish thee for thy arrogance.
</p>
<p>  <cite>Henry Spencer&#8217;s 10 Commandments for C Programmers</cite>
</p></blockquote>
<p>In the olden days, before Exceptions were invented, we wanted to write code like this:</p>
<pre><code>// Called when the user presses the "triple" button
void onTripleButtonPressed() {
  String valueEntered = readValueFromField();
  int valueToDisplay = triple(stringToInteger(valueEntered));
  displayResult(integerToString(valueToDisplay));
}

int triple(int value) {
  return value * 3;
}
</code></pre>
<p>Unfortunately, we were forced to write code like this instead:</p>
<pre><code>int onTripleButtonPressed() {
  StringHolder valueEntered = new StringHolder();
  int resultCode = readValueFromField(valueEntered);
  if (isError(resultCode))
    return resultCode;

  IntHolder valueAsInteger = new IntHolder();
  resultCode = stringToInteger(valueAsInteger, valueEntered.value);
  if (isError(resultCode))
    return resultCode;

  resultCode = triple(valueAsInteger);
  if (isError(resultCode))
    return resultCode;

  StringHolder valueToDisplay = new StringHolder();
  resultCode = integerToString(valueToDisplay, valueAsInteger.value);
  if (isError(resultCode))
    return resultCode;

  resultCode = displayResult(valueToDisplay.value);
  if (isError(resultCode))
    return resultCode;

  return OK; // :-)
}

private int triple(IntHolder valueAsInteger) {
  int result = valueAsInteger.value * 3;
  if (isOverflow(result)) 
    return ERROR_OVERFLOW;
  valueAsInteger.value = result;
  return OK;
}
</code></pre>
<p>Yes, we were FORCED to check the result of each and every operation.  There was no other way to write reliable software.  As you can see:</p>
<ul>
<li>Code size is more than three times</li>
<li>Logic becomes obscure</li>
<li>Functions are forced to return two values; the intended result AND an error code.</li>
<li>You always, always, always had to return an error code from all functions.</li>
</ul>
<p>There was no alternative, until exceptions came along.  Then we were finally able to write simple code in a reliable way:</p>
<pre><code>// Called when the user presses the "triple" button
// Exceptions are handled here
void onTripleButtonPressed() {
  try {
    tryOnTripleButtonPressed();
  } catch (Exception e) {
    alertUser(e.getMessage());
  }
}

// Business logic is handled here
private void tryOnTripleButtonPressed() {
  String valueEntered = readValueFromField();
  int valueToDisplay = triple(stringToInteger(valueEntered));
  displayResult(integerToString(valueToDisplay));
}
</code></pre>
<p>In this example, all exception handling is centralized <strong>at the point where the GUI gives control to our code</strong>.  The code that performs what the user really wanted is in another function, which contains exactly the same clean code of the first example.</p>
<p>Thus, the invention of exception handling allows us to cleanly separate code that performs the happy path, and code that handles the many possible exceptional conditions: integer overflow, I/O exceptions, GUI widgets being improperly configured, etc.  </p>
<p>What you saw in the previous example is the </p>
<blockquote><p>
<strong>Fundamental Pattern of Exception Handling</strong>: centralize exception handling at the point where the GUI gives control to our code.  No other exception handling should appear anywhere else.
</p></blockquote>
<p>It turns out that the Fundamental Pattern of Exception Handling is the only pattern that we need to know.  There are other cases where we are tempted to write a try-catch, but it turns out that we nearly always have better ways to do it.</p>
<h3 id="antipattern_nostalgia_for_the_bad_old_days">Antipattern: <strong>nostalgia</strong> (for the bad old days)</h3>
<p>There is an ineffective style of coding that we sometimes see in legacy code:</p>
<pre><code>void onTripleButtonPressed() {
  String valueEntered = null;
  try {
    valueEntered = readValueFromField();
  } catch (Exception e) {
    logger.log(ERROR, "can't read from field", e);
    return;
  }

  int valueEnteredAsInteger = 0;
  try {
    valueEnteredAsInteger = Integer.parseInt(valueEntered);
  } catch (Exception e) {
    logger.log(ERROR, "parse exception", e);
    return;
  }

  int valueToDisplay = 0;
  try {
    valueToDisplay = triple(valueEnteredAsInteger);
  } catch (Exception e) {
    logger.log(ERROR, "overflow", e);
    return;
  }

  try {
    displayResult(integerToString(valueToDisplay));
  } catch (Exception e) {
    logger.log(ERROR, "something went wrong", e);
    return;
  }
}
</code></pre>
<p>As you can see, this is just as bad as the olden days code!  But while in the old times we had no alternatives, now we do.  Just use exceptions the way they were intended to.</p>
<h3 id="stay_up_no_matter_what">Stay up no matter what</h3>
<p>Sometimes there is a feeling that we should write code that &#8220;stays up no matter what happens&#8221;.  For instance:</p>
<pre><code>private void tryOnTripleButtonPressed() {
  String valueEntered = readValueFromField();
  int valueToDisplay = triple(stringToInteger(valueEntered));

  try {
    sendEmail(userEmail, "You tripled " + valueEntered);
  } catch (Exception e) {
    logger.log(WARNING, "could not send email", e);
    // continue
  }

  displayResult(integerToString(valueToDisplay));
}
</code></pre>
<p>Here we have added an email confirmation; whenever the user presses the button, they will also receive an email.  Now we should ask ourselves which of the two: </p>
<ol style="list-style-type: decimal">
<li>is the sending of the email an integral and fundamental part of what should happen when the user presses the button?  </li>
<li>Or is it an accessory part should never stop the completion of the rest of the action?</li>
</ol>
<p>This is a business decision.  If the business decides on 1., then we should remove the try-catch.  By applying the Fundamental Pattern, proper logging of the exception will be done elsewhere.  We end up with:</p>
<pre><code>private void tryOnTripleButtonPressed() {
  String valueEntered = readValueFromField();
  int valueToDisplay = triple(stringToInteger(valueEntered));
  sendEmail(userEmail, "You tripled " + valueEntered);
  displayResult(integerToString(valueToDisplay));
}
</code></pre>
<p>Clean code again.  Yay!</p>
<p>If the business decides on 2., then we cannot allow sendMail to throw exceptions that might stop the processing of the user action.  What do we do now?  Do we have to keep the try-catch?</p>
<p>The answer is yes, but out of the way.  There are two ways to do this: the easy way and the simple way.  If you do it the <strong>easy</strong> way, you will move the try-catch <em>inside</em> the sendMail function.  You will get code like this:</p>
<pre><code>private void tryOnTripleButtonPressed() {
    String valueEntered = readValueFromField();
    int valueToDisplay = triple(stringToInteger(valueEntered));
    sendMail(userEmail, "You tripled " + valueEntered);     
    displayResult(integerToString(valueToDisplay));
  }

private void sendMail(EmailAddress email, String message) {
  try {
      trySendEmail(email, message);
  } catch (Exception e) {
    logger.log(WARNING, "could not send email", e);
    // continue
  }
}
</code></pre>
<p>You have moved exception handling for sending mail to a dedicated function (good).  However, you still have complicated code tightly coupled to the user action.  The sending of email, which is an <em>accessory</em> operation, makes understanding the fundamental operation more difficult (bad!)</p>
<p>What is the <strong>simple</strong> way, then?  In the simple way we eliminate the tight coupling betweeen tripling the number and sending the email.  There are several patterns that we might use; a typical choice would be &#8220;publish-subscribe&#8221;.  We set up a subscriber that waits for the &#8220;User Pressed the Triple Button&#8221; event.  In the main onTripleButtonPressed function we don&#8217;t know nor care what the subscriber does.  It might send email, write logs, compute statistics, or maybe distribute the event to a list of other subscribers.  We don&#8217;t know nor care!  The code looks like this:</p>
<pre><code>private void tryOnTripleButtonPressed() {
  String valueEntered = readValueFromField();
  int valueToDisplay = triple(stringToInteger(valueEntered));
  subscriber.notifyTripled(valueEntered);
  displayResult(integerToString(valueToDisplay));
}
</code></pre>
<p>In the Simple way, the Fundamental Pattern has been respected: for the subscriber object, the start of the processing is within the notifyTripled method.  We have clean <em>and</em> loosely coupled code.</p>
<h3 id="avoid_resource_leaking">Avoid resource leaking</h3>
<p>Another time when we are tempted to use a try-catch in violation of the Fundamental Pattern is when we open some resource that must be closed.</p>
<pre><code>void doSomethingImportant() {
  Reader reader = null;
  try {
    reader = new FileReader("foo bar");
    doSomethingWith(reader);
    reader.close();
  } catch (Exception e) {
    if (null != reader) {
      reader.close();
    }
    // report the exception to the callers
    throw new RuntimeException(e);
  } 
}
</code></pre>
<p>It is correct that we close the reader in the event that something within doSomethingWith throws an exception. But we don&#8217;t want the catch; a finally clause is better:</p>
<pre><code>void doSomethingImportant() throws IOException {
  Reader reader = null;
  try {
    reader = new FileReader("foo bar");
    doSomethingWith(reader);
  } finally {
    if (null != reader)
      reader.close();
  } 
}
</code></pre>
<p>This way, we don&#8217;t even need to worry about rethrowing the exception to our callers.  The code is both correct <em>and</em> clear.</p>
<h3>Irritating APIs</h3>
<p>Some APIs force us to use more try-catches that we&#8217;d like. An example in Java is how to read data from a database using JDBC:</p>
<pre><code>PreparedStatement statement = connection.prepareStatement("...");
try {
    ResultSet results = statement.executeQuery();
    try {
        while (results.next()) {
            // ...
        }
    } finally {
        results.close();
    }
} finally {
    statement.close();
} 
</code></pre>
<p>Here the problem lies in how the JDBC API is defined.  We can&#8217;t change it of course, but we should encapsulate all uses of the JDBC API in a single class, so that we don&#8217;t have to look at this code anymore.  Treat that class as your adapter to JDBC.  For instance:</p>
<pre><code>Results results = new Database(connection).read("select * from foo");
  // use results
</code></pre>
<p>Of course our object is a lot less flexible than the JDBC API.  This is expected: we want to decide <em>how</em> we use JDBC and encapsulate that &#8220;how&#8221; in a single place.</p>

				
				<p class="postmetadata alt">
					<small>
						This entry was posted
												on 20 November 2013 at 12:50						and is filed under <a href="category/agile.html" rel="category tag">Agile</a>.
						You can follow any responses to this entry through the <a href="875/feed">RSS 2.0</a> feed.

													You can <a href="#respond">leave a response</a>, or <a href="875/trackback" rel="trackback">trackback</a> from your own site.

						
					</small>
				</p>

			</div>
		</div>

	
<!-- You can start editing here. -->

	<h3 id="comments">6 Responses to &#8220;Notes on exception handling&#8221;</h3> 

	<ol class="commentlist">

	
		<li class="alt" id="comment-96356">
			<cite>Philip Schwarz</cite> Says:
						<br />

			<small class="commentmetadata"><a href="#comment-96356" title="">November 24th, 2013 at 08:07</a> </small>

			<p>Hi Matteo. </p>
<p>Interesting post, especially the usage of the publish-subscribe pattern to eliminate the tight coupling between tripling the number and sending the email.</p>
<p>I have a question about the following though:</p>
<p>void onTripleButtonPressed() {<br />
  try {<br />
    tryOnTripleButtonPressed();<br />
  } catch (Exception e) {<br />
    alertUser(e.getMessage());<br />
  }<br />
}</p>
<p>Normally when I see code catching java.lang.Exception, I bring up the following two reasons why I think it should not be done.</p>
<p>The first reason is explained in the Catching Exception antipattern:<br />
<a href="https://today.java.net/pub/a/today/2006/04/06/exception-handling-antipatterns.html#catchingException" rel="nofollow">https://today.java.net/pub/a/today/2006/04/06/exception-handling-antipatterns.html#catchingException</a></p>
<p>My other reason why catching java.lang.Exception (or java.lang.Throwable for that matter) is an antipattern is that one of the subclasses of Exception is RuntimeException (<a href="http://javajazzle.files.wordpress.com/2011/02/hierarchy_of_java_exceptions.png" rel="nofollow">http://javajazzle.files.wordpress.com/2011/02/hierarchy_of_java_exceptions.png</a>).  So whenever you catch Exception, you are also catching RuntimeException. But runtime exceptions are generally supposed to indicate programming errors. </p>
<p>A correctly working system is generally not expected to encounter runtime exceptions. When a system encounters a runtime exception, you wan to be sure that the system informs you about it, and you then want to go and fix the programming error that caused it. </p>
<p>There is no point in handling a runtime exception programmatically (other than reporting it) because if it is ever encountered, you will fix the error that caused it, so it should never be raised again. If you catch Exception, you run the risk of handling runtime exceptions in the same way as checked exceptions, which is almost certainly incorrect. </p>
<p>So if you really have to catch Exception (this is highly discouraged), you should first catch RuntimeException and rethrow it:</p>
<p>catch(RuntimeException re)<br />
{<br />
    throw re<br />
}<br />
catch(Exception e)<br />
{<br />
    &#8230;some logic&#8230;<br />
}</p>
<p>Of course in most systems that are in production, instead of allowing a runtime exception to bubble all the way up, which would bring the system to a halt, we want to catch them at the top of our exception handling structure, and inform interested parties about it.</p>
<p>If we apply the Fundamental Pattern of Exception Handling, there is only one point where exceptions are handled, so it makes sense to catch runtime exceptions as well, but we want to handle them differently from checked exceptions. I guess pointing this out was out of scope for your post.</p>
<p>What are your thoughts on the two reasons I mentioned for not catching Exception?</p>

		</li>

	
	
		<li class="" id="comment-96359">
			<cite>matteo</cite> Says:
						<br />

			<small class="commentmetadata"><a href="#comment-96359" title="">November 25th, 2013 at 10:25</a> </small>

			<p>Hello Philip,</p>
<p>thanks for your comment.  I don&#8217;t share your concern about catching Exception.  RuntimeExceptions might happen for all kinds of reasons, like NullPointerExceptions which are certainly due to logic errors, but also OutOfMemoryError or even applicative errors that the developer chose to declare as unchecked exceptions.  </p>
<p>I think they should be handled uniformly; whenever an exception is thrown, the fact is that we cannot do what the user asked us to do.  We log appropriately and present an appropriate error message to the user.  Even logic errors are a fact of life; I want the system to handle them as gracefully as possible.</p>
<p>Exceptions should be exceptional; for instance in my example above, there should probably be a check that the string actually contains a parseable integer, so that no ParseException is ever thrown.  Most exceptions should be prevented by checking that an operation can be performed before we perform it.  If we do this, then anything that is thrown should either be something that we didn&#8217;t think of, like a hole in our validation checks, or something that we cannot possibly do anything about, like our database server is unreachable or we&#8217;re out of memory.  I don&#8217;t think you can gain much by catching exceptions selectively then.</p>
<p>By the way: the pattern that I called the &#8220;Fundamental Pattern of Exception Handling&#8221; is not something I invented, of course.  I learned about it when I worked with the MacApp application framework, circa 1990 :-)</p>

		</li>

	
	
		<li class="alt" id="comment-96365">
			<cite>Uberto</cite> Says:
						<br />

			<small class="commentmetadata"><a href="#comment-96365" title="">November 26th, 2013 at 17:59</a> </small>

			<p>Hi Matteo, while I certainly agree with you here, I think you are considering only the &#8220;easy&#8221; case.<br />
In all examples here, exception are raised for &#8220;exceptional&#8221; cases, either bugs in our code or a failure of our environment (mail server down).<br />
I&#8217;ll wait until you wrote something about exceptions used for flow control.  (<a href="http://c2.com/cgi/wiki?DontUseExceptionsForFlowControl" rel="nofollow">http://c2.com/cgi/wiki?DontUseExceptionsForFlowControl</a>)</p>
<p>:)</p>

		</li>

	
	
		<li class="" id="comment-96407">
			<cite>Philip Schwarz</cite> Says:
						<br />

			<small class="commentmetadata"><a href="#comment-96407" title="">November 30th, 2013 at 13:27</a> </small>

			<p>&gt;RuntimeExceptions might happen for all kinds of reasons, like NullPointerExceptions which are certainly due to logic errors, but also OutOfMemoryError or even applicative errors </p>
<p>No RuntimException can happen due to OutOfMemoryError because the latter extends Error:</p>
<p>java.lang.Object &lt;== java.lang.Throwable  &lt;== java.lang.Error &lt;==  java.lang.VirtualMachineError &lt;== java.lang.OutOfMemoryError.</p>
<p>So when you catch Exception, you run no risk of catching any Error.</p>
<p>While RuntimeException should not normally be caught because it indicates a programming error, Error should not normally be caught because it is a serious problem and most of the time and  abnormal condition.</p>
<p>Philip</p>

		</li>

	
	
		<li class="alt" id="comment-96409">
			<cite>Philip Schwarz</cite> Says:
						<br />

			<small class="commentmetadata"><a href="#comment-96409" title="">November 30th, 2013 at 20:38</a> </small>

			<p>Willie Wheeler, author of &#8216;Spring in Practice&#8217; has a name for the second concern I have with catching Exception: &#8216;Mask via broad catch, and misreport&#8217; (see Wheeler&#8217;s answer to StackOverflow question: <a href="http://stackoverflow.com/questions/6591470/checked-vs-unchecked-exceptions-in-service-layer" rel="nofollow">http://stackoverflow.com/questions/6591470/checked-vs-unchecked-exceptions-in-service-layer</a>).</p>
<p>Philip</p>

		</li>

	
	
		<li class="" id="comment-96411">
			<cite>Philip Schwarz</cite> Says:
						<br />

			<small class="commentmetadata"><a href="#comment-96411" title="">November 30th, 2013 at 21:48</a> </small>

			<p>One way in which Nat Pryce, author of Growing Object Oriented Software &#8211; Guided by Tests, stressed  the kind of error that RuntimeExceptions are designed to report, was by blogging (<a href="http://nat.truemesh.com/archives/000643.html" rel="nofollow">http://nat.truemesh.com/archives/000643.html</a>) that RuntimeException be renamed to StupidProgrammerException. Another way he did that was by suggesting (<a href="http://www.natpryce.com/articles/000739.html" rel="nofollow">http://www.natpryce.com/articles/000739.html</a>) that checked exceptions due to programming errors be converted in a dedicated Defect exception extending RuntimeException.</p>

		</li>

	
	
	</ol>

 


<h3 id="respond">Leave a Reply</h3>


<form action="http://matteo.vaccari.name/blog/wp-comments-post.php" method="post" id="commentform">


<p><input type="text" name="author" id="author" value="" size="22" tabindex="1" />
<label for="author"><small>Name (required)</small></label></p>

<p><input type="text" name="email" id="email" value="" size="22" tabindex="2" />
<label for="email"><small>Mail (will not be published) (required)</small></label></p>

<p><input type="text" name="url" id="url" value="" size="22" tabindex="3" />
<label for="url"><small>Website</small></label></p>


<!--<p><small><strong>XHTML:</strong> You can use these tags: &lt;a href=&quot;&quot; title=&quot;&quot;&gt; &lt;abbr title=&quot;&quot;&gt; &lt;acronym title=&quot;&quot;&gt; &lt;b&gt; &lt;blockquote cite=&quot;&quot;&gt; &lt;cite&gt; &lt;code&gt; &lt;del datetime=&quot;&quot;&gt; &lt;em&gt; &lt;i&gt; &lt;q cite=&quot;&quot;&gt; &lt;s&gt; &lt;strike&gt; &lt;strong&gt; </small></p>-->

<p><textarea name="comment" id="comment" cols="100%" rows="10" tabindex="4"></textarea></p>

<p><input name="submit" type="submit" id="submit" tabindex="5" value="Submit Comment" />
<input type="hidden" name="comment_post_ID" value="875" />
</p>
<p style="display: none;"><input type="hidden" id="akismet_comment_nonce" name="akismet_comment_nonce" value="5afb626c77" /></p><p style="display: none;"><input type="hidden" id="ak_js" name="ak_js" value="67"/></p>
</form>



	
    </div>

    <div id='right-col'>
        <div id="sidebar">
    <ul>

      <li>
        <form method="get" id="searchform" action="http://matteo.vaccari.name/blog/">
  <div>
    <input tabindex="1" type="text" value="" name="s" id="s" />
    <input type="submit" id="searchsubmit" value="Search" />
  </div>
</form>
      </li>

      <li>

        <h2>About</h2>

	I am a software developer working with <a href="http://thoughtworks.com/">Thoughtworks Italia</a>.
        I'm a fan of <a
        href="http://extremeprogramming.org/">Extreme Programming</a>.
        I&nbsp;helped organize the <a
        href="http://essap.dicom.uninsubria.it/">European Summer School on
        Agile Programming</a>

      </li>

      <li>
            </li>

      <li class="pagenav"><h2>Pages</h2><ul><li class="page_item page-item-851"><a href="../selected-tdd-resources-to-start.html">Selected TDD Resources To Start</a></li>
<li class="page_item page-item-583"><a href="../tdd-resources.html">TDD Resources</a></li>
<li class="page_item page-item-247"><a href="../the-hexagonal-architecture.html">The Hexagonal Architecture</a></li>
</ul></li>

      
      <li><h2>Archives</h2>
        <ul>
        	<li><a href='date/2018/10.html'>October 2018</a></li>
	<li><a href='date/2017/09.html'>September 2017</a></li>
	<li><a href='date/2016/08.html'>August 2016</a></li>
	<li><a href='date/2016/03.html'>March 2016</a></li>
	<li><a href='date/2016/01.html'>January 2016</a></li>
	<li><a href='date/2014/12.html'>December 2014</a></li>
	<li><a href='date/2014/10.html'>October 2014</a></li>
	<li><a href='date/2014/06.html'>June 2014</a></li>
	<li><a href='date/2014/02.html'>February 2014</a></li>
	<li><a href='date/2013/12.html'>December 2013</a></li>
	<li><a href='date/2013/11.html'>November 2013</a></li>
	<li><a href='date/2013/06.html'>June 2013</a></li>
	<li><a href='date/2013/05.html'>May 2013</a></li>
	<li><a href='date/2013/03.html'>March 2013</a></li>
	<li><a href='date/2013/02.html'>February 2013</a></li>
	<li><a href='date/2012/10.html'>October 2012</a></li>
	<li><a href='date/2012/08.html'>August 2012</a></li>
	<li><a href='date/2012/07.html'>July 2012</a></li>
	<li><a href='date/2012/06.html'>June 2012</a></li>
	<li><a href='date/2012/05.html'>May 2012</a></li>
	<li><a href='date/2012/04.html'>April 2012</a></li>
	<li><a href='date/2012/01.html'>January 2012</a></li>
	<li><a href='date/2011/12.html'>December 2011</a></li>
	<li><a href='date/2011/11.html'>November 2011</a></li>
	<li><a href='date/2011/10.html'>October 2011</a></li>
	<li><a href='date/2011/07.html'>July 2011</a></li>
	<li><a href='date/2011/06.html'>June 2011</a></li>
	<li><a href='date/2011/05.html'>May 2011</a></li>
	<li><a href='date/2011/04.html'>April 2011</a></li>
	<li><a href='date/2011/03.html'>March 2011</a></li>
	<li><a href='date/2011/02.html'>February 2011</a></li>
	<li><a href='date/2010/11.html'>November 2010</a></li>
	<li><a href='date/2010/10.html'>October 2010</a></li>
	<li><a href='date/2010/09.html'>September 2010</a></li>
	<li><a href='date/2010/08.html'>August 2010</a></li>
	<li><a href='date/2010/07.html'>July 2010</a></li>
	<li><a href='date/2010/06.html'>June 2010</a></li>
	<li><a href='date/2010/05.html'>May 2010</a></li>
	<li><a href='date/2010/04.html'>April 2010</a></li>
	<li><a href='date/2010/03.html'>March 2010</a></li>
	<li><a href='date/2010/02.html'>February 2010</a></li>
	<li><a href='date/2010/01.html'>January 2010</a></li>
	<li><a href='date/2009/12.html'>December 2009</a></li>
	<li><a href='date/2009/11.html'>November 2009</a></li>
	<li><a href='date/2009/10.html'>October 2009</a></li>
	<li><a href='date/2009/09.html'>September 2009</a></li>
	<li><a href='date/2009/07.html'>July 2009</a></li>
	<li><a href='date/2009/05.html'>May 2009</a></li>
	<li><a href='date/2009/04.html'>April 2009</a></li>
	<li><a href='date/2009/03.html'>March 2009</a></li>
	<li><a href='date/2009/02.html'>February 2009</a></li>
	<li><a href='date/2009/01.html'>January 2009</a></li>
	<li><a href='date/2008/12.html'>December 2008</a></li>
	<li><a href='date/2008/11.html'>November 2008</a></li>
	<li><a href='date/2008/10.html'>October 2008</a></li>
	<li><a href='date/2008/09.html'>September 2008</a></li>
	<li><a href='date/2008/07.html'>July 2008</a></li>
	<li><a href='date/2008/06.html'>June 2008</a></li>
	<li><a href='date/2008/05.html'>May 2008</a></li>
	<li><a href='date/2008/04.html'>April 2008</a></li>
	<li><a href='date/2008/01.html'>January 2008</a></li>
	<li><a href='date/2007/11.html'>November 2007</a></li>
	<li><a href='date/2007/10.html'>October 2007</a></li>
	<li><a href='date/2007/09.html'>September 2007</a></li>
	<li><a href='date/2007/06.html'>June 2007</a></li>
	<li><a href='date/2007/05.html'>May 2007</a></li>
	<li><a href='date/2007/04.html'>April 2007</a></li>
	<li><a href='date/2007/03.html'>March 2007</a></li>
	<li><a href='date/2007/02.html'>February 2007</a></li>
	<li><a href='date/2007/01.html'>January 2007</a></li>
	<li><a href='date/2006/12.html'>December 2006</a></li>
	<li><a href='date/2006/11.html'>November 2006</a></li>
	<li><a href='date/2006/10.html'>October 2006</a></li>
	<li><a href='date/2006/09.html'>September 2006</a></li>
	<li><a href='date/2006/08.html'>August 2006</a></li>
	<li><a href='date/2006/07.html'>July 2006</a></li>
	<li><a href='date/2006/06.html'>June 2006</a></li>
	<li><a href='date/2006/05.html'>May 2006</a></li>
	<li><a href='date/2006/04.html'>April 2006</a></li>
	<li><a href='date/2006/03.html'>March 2006</a></li>
	<li><a href='date/2006/02.html'>February 2006</a></li>
	<li><a href='date/2006/01.html'>January 2006</a></li>
        </ul>
      </li>

      <li><h2>Categories</h2>
        <ul>
        	<li class="cat-item cat-item-1"><a href="category/agile.html" >Agile</a> (229)
</li>
	<li class="cat-item cat-item-3"><a href="category/books.html" >Books</a> (2)
</li>
	<li class="cat-item cat-item-11"><a href="category/design-exercise.html" >Design Exercise</a> (5)
</li>
	<li class="cat-item cat-item-6"><a href="category/essay.html" >Essay</a> (13)
</li>
	<li class="cat-item cat-item-8"><a href="category/expressive-code.html" >Expressive Code</a> (11)
</li>
	<li class="cat-item cat-item-10"><a href="category/fundamentals.html" >Fundamentals</a> (6)
</li>
	<li class="cat-item cat-item-9"><a href="category/other.html" >other</a> (1)
</li>
        </ul>
      </li>

      
    </ul>
  </div>

    </div>
</div>


<hr />
<div id="footer">
<address>
  <script type="text/javascript">eval(unescape('%64%6f%63%75%6d%65%6e%74%2e%77%72%69%74%65%28%27%3c%61%20%68%72%65%66%3d%22%6d%61%69%6c%74%6f%3a%76%61%63%63%61%72%69%40%70%6f%62%6f%78%2e%63%6f%6d%22%3e%76%61%63%63%61%72%69%40%70%6f%62%6f%78%2e%63%6f%6d%3c%2f%61%3e%27%29%3b'))</script>
</address>
</div>
</div>
  <script type='text/javascript' src='../wp-includes/js/wp-embed.min3c21.js?ver=5.1.1'></script>
<script async="async" type='text/javascript' src='../wp-content/plugins/akismet/_inc/form0235.js?ver=4.1.1'></script>
</body>

<!-- Mirrored from matteo.vaccari.name/blog/archives/875 by HTTrack Website Copier/3.x [XR&CO'2014], Fri, 24 Sep 2021 10:25:15 GMT -->
</html>
